File: src/cpp/Ca_HVA.cpp
#include "Ca_HVA.hpp"

#include <cmath>

namespace {
const scalar_t e = 2.718281828459045235360287471352;
const scalar_t gCa_HVAbar = 0.00001;
}

std::pair<scalar_t, scalar_t> Ca_HVA_derivatives(std::pair<scalar_t, scalar_t> state, scalar_t V) {
    if (V == -27) {
        V += 0.0001;
    }

    const scalar_t mAlpha = -(0.055 * (V + 27)) / (std::pow(e, (-(V + 27) / 3.8)) - 1);
    const scalar_t mBeta = 0.94 * std::pow(e, (-(V + 75) / 17));
    const scalar_t mInf = mAlpha / (mAlpha + mBeta);
    const scalar_t mTau = 1.0 / (mAlpha + mBeta);

    const scalar_t hAlpha = 0.000457 * std::pow(e, (-(V + 13) / 50));
    const scalar_t hBeta = 0.0065 / (std::pow(e, (-(V + 15) / 28)) + 1);
    const scalar_t hInf = hAlpha / (hAlpha + hBeta);
    const scalar_t hTau = 1.0 / (hAlpha + hBeta);

    const scalar_t m = state.first;
    const scalar_t h = state.second;

    const scalar_t dm_dt = (mInf - m) / mTau;
    const scalar_t dh_dt = (hInf - h) / hTau;

    return std::make_pair(dm_dt, dh_dt);
}

scalar_t Ca_HVA_current(std::pair<scalar_t, scalar_t> state, scalar_t V, scalar_t E_Ca, scalar_t dt) {
    if (V == -27) {
        V += 0.0001;
    }

    const scalar_t mAlpha = -(0.055 * (V + 27)) / (std::pow(e, (-(V + 27) / 3.8)) - 1);
    const scalar_t mBeta = 0.94 * std::pow(e, (-(V + 75) / 17));
    const scalar_t mInf = mAlpha / (mAlpha + mBeta);
    const scalar_t mTau = 1.0 / (mAlpha + mBeta);

    const scalar_t hAlpha = 0.000457 * std::pow(e, (-(V + 13) / 50));
    const scalar_t hBeta = 0.0065 / (std::pow(e, (-(V + 15) / 28)) + 1);
    const scalar_t hInf = hAlpha / (hAlpha + hBeta);
    const scalar_t hTau = 1.0 / (hAlpha + hBeta);

    scalar_t m = state.first;
    scalar_t h = state.second;

    m = mInf + (m - mInf) * std::pow(e, (-dt / mTau));
    h = hInf + (h - hInf) * std::pow(e, (-dt / hTau));

    const scalar_t gCa_HVAst = gCa_HVAbar * m * m * h;
    const scalar_t I_Ca = gCa_HVAst * (V - E_Ca);
    return I_Ca;
}

File: src/conductances/Ca_HVA.py
"""
Ca_HVA: High-voltage-activated calcium conductance

Extracted from BBP NEURON mechanism: Ca_HVA.mod
Neuron type: pyramidal neuron

State variables:
    m: activation gate
    h: inactivation gate
"""

# euler's constant to 30 digits of precision
e = 2.718281828459045235360287471352

# maximal peak conductance density of channel per unit membrane area (S/cm^2)
gCa_HVAbar = 0.00001


def derivatives(state, V):
    """
    Compute steady-state values and time constants,
    then compute time derivatives of the gating variables.

    Inputs:
        - state (tuple of floats): (m, h) gating variables
        - V (float): membrane potential (mV)

    Returns: 
        - dm_dt (float): time derivative of m
        - dh_dt (float): float time derivative of h
    """

    # voltage shift
    if V == -27:
        V += 0.0001
    
    # steady state activation and activation time constant (ms)
    mAlpha =  -(0.055 * (V + 27)) / (e**(-(V + 27) / 3.8) - 1)        
    mBeta  =  (0.94* e**(-(V + 75) / 17))
    mInf = mAlpha / (mAlpha + mBeta)
    mTau = 1 / (mAlpha + mBeta)
    
    # steady state inactivation and inactivation time constant (ms)
    hAlpha =  (0.000457 * e**(-(V + 13) / 50))
    hBeta  =  (0.0065 / (e**(-(V + 15) / 28) + 1))
    hInf = hAlpha / (hAlpha + hBeta)
    hTau = 1 / (hAlpha + hBeta)
    
    m, h = state

    dm_dt = (mInf - m) / mTau
    dh_dt = (hInf - h) / hTau

    return dm_dt, dh_dt

def current(state, V, E_Ca, dt=0.025):
    """
    Compute current based on gating variables and membrane potential.

    Inputs:
        - state (tuple of floats): (m, h) gating variables
        - V (float): membrane potential (mV)
        - E_Ca (float): calcium reversal potential (mV)
        - time step (float): time step of derivative measurement

    Returns: 
        - I_Ca (float): current (mA/cm^2)
    """
    m, h = state
    
    # voltage shift
    if V == -27:
        V += 0.0001
    
    # steady state activation and activation time constant (ms)
    mAlpha =  -(0.055 * (V + 27)) / (e**(-(V + 27) / 3.8) - 1)        
    mBeta  =  (0.94* e**(-(V + 75) / 17))
    mInf = mAlpha / (mAlpha + mBeta)
    mTau = 1 / (mAlpha + mBeta)
    
    # steady state inactivation and inactivation time constant (ms)
    hAlpha =  (0.000457 * e**(-(V + 13) / 50))
    hBeta  =  (0.0065 / (e**(-(V + 15) / 28) + 1))
    hInf = hAlpha / (hAlpha + hBeta)
    hTau = 1 / (hAlpha + hBeta)
    
    # cnexp update
    m = mInf + (m - mInf) * e**(-dt/mTau)
    h = hInf + (h - hInf) * e**(-dt/hTau)
    
    gCa_HVAst = gCa_HVAbar * m * m * h
    I_Ca = gCa_HVAst * (V - E_Ca)
    return I_Ca